### ISSUE TITLE
Unsolicited iOS14 clipboard permission prompt

### ISSUE THREAD

**Issue by justinmc** [flutter/flutter#60145]
At WWDC2020, Apple announced that it was cracking down on clipboard sniffing by showing a prompt when an app attempts to access the clipboard.  This likely means that Flutter apps that feature text input will trigger this at strange times.

As of https://github.com/flutter/flutter/pull/54902, Flutter began checking the clipboard contents to decide whether or not to show the paste button in the text selection menu.  Previously, it would always show the button, even when nothing pasteable existed in the clipboard.

When exactly are Flutter apps seeing this clipboard permission dialog?  What should we do about it?

Some options:

  1. Revert to the old behavior (show the paste button regardless of clipboard contents).
  1. Revert only for iOS, or only or iOS 14.  Provide the developer an option to enable the feature on iOS 14 anyway?

**Comment by EliasDeuss** [comment]
Yea I have it being spammed in one of my apps.

![RPReplay-Final1592862237](https://user-images.githubusercontent.com/13773566/85482580-06064a80-b578-11ea-949d-78db963e2b6c.gif)

**Comment by justinmc** [comment]
@EliasDeuss Thanks for posting the gif!  I hadn't actually reproduced it yet.  Looks bad, we'll definitely have to fix this.

**Comment by EliasDeuss** [comment]
> @EliasDeuss Thanks for posting the gif!  I hadn't actually reproduced it yet.  Looks bad, we'll definitely have to fix this.

It also might be that bad because the page has a StreamBuilder on it. (Don’t know if that would effect it.)

**Comment by tamirrab** [comment]
Every flutter app including apps on the stores have this “paste notification”
It is very bad,
Customer think we are storing there information

**Comment by goderbauer** [comment]
Apparently, Apple is also introducing a set of new APIs for clipboards: https://9to5mac.com/2020/06/24/ios-14-clipboard-notifications/

I haven't dived into it, but maybe those can help us to determine if there's anything interesting in the clipboard before actually accessing it.

To people who have seen the notification in the wild for a Flutter app: When does it actually show up?

**Comment by goderbauer** [comment]
We should fix this before iOS 14 launches.

**Comment by goderbauer** [comment]
@jonahwilliams also pointed to https://developer.apple.com/documentation/uikit/uipasteboard/1829416-hasstrings?language=objc, which may help us.

**Comment by EliasDeuss** [comment]
> @jonahwilliams also pointed to https://developer.apple.com/documentation/uikit/uipasteboard/1829416-hasstrings?language=objc, which may help us. 

@justinmc Have a look at this.

**Comment by justinmc** [comment]
As a stopgap solution, I propose we disable the automatic hiding/showing of the paste button on iOS right now: https://github.com/flutter/flutter/pull/60316

This will eliminate the notification from showing up during text composition due to passive checking of the clipboard.  From there, it seems like hasStrings (linked above) would be a great solution for allowing us to re-enable the paste button hiding behavior without triggering the notification.

**Comment by justinmc** [comment]
However, neither my PR nor use of hasStrings will eliminate the notification when actually pressing the paste button!  This could be a major annoyance for Flutter apps.  Is there any way to work around this that I'm not thinking of?

**Comment by justinmc** [comment]
@goderbauer You were right, I just tried a native SwiftUI app running on an iOS 14 emulator with nothing but a text field, and it does show the notification!  That's good news for Flutter.

**Comment by goderbauer** [comment]
Awesome. So with the help of `hasStrings` we should be able to implement the native behavior in Flutter where we only show the Paste button if there is something pastable and the notification only triggers when the paste button is pressed?

**Comment by justinmc** [comment]
That's right (though to be specific, I think it shows when the copy button is pressed).

**Comment by mehmetf** [comment]
If I am parsing this thread correctly, the symptom (displaying lots of clipboard prompts) should now be fixed in master but we are keeping this open to properly solve it with `hasStrings`?

**Comment by justinmc** [comment]
That's right.  The downside of the current solution is that we can't hide the paste button on iOS when the clipboard has nothing to paste.  Using `hasStrings` will allow us to hide the paste button again without triggering the clipboard prompt.

**Comment by xster** [comment]
@justinmc can you assign an expected milestone for this issue?

**Comment by justinmc** [comment]
Done, thanks for the reminder 👍

**Comment by csongorkeller** [comment]
Is it already part of any flutter version? if yes, how can I use hasString?

**Comment by justinmc** [comment]
It is not yet, see the two PRs mentioned above (https://github.com/flutter/engine/pull/19859 and https://github.com/flutter/flutter/pull/61774).

Once they are merged, hasStrings will be used internally by Flutter automatically.  If you want to use it yourself too, I may end up adding it as a static method on Clipboard.

**Comment by csongorkeller** [comment]
Any news on this topic? As I see there is a merged PR, is it already part of any flutter version? Thanks in advance

**Comment by justinmc** [comment]
@csongorkeller A quick fix was merged for this (https://github.com/flutter/flutter/pull/60316), but I'm keeping this issue open while I work on a cleaner fix.  On all current Flutter channels, there should be no unsolicited clipboard permission prompts.

**Comment by csongorkeller** [comment]
Awesome. Thanks. So this means if I use the last recent flutter version on master, then It'll be fixed for now?

**Comment by justinmc** [comment]
Yes that's right 👍   The quick fix is in v1.22.0-1.0.pre and above.

**Comment by csongorkeller** [comment]
Tried with those versions, but I never reached the point to test the clipboard permission prompt issue coz 1.22 slowed my app down significantly. :(

**Comment by justinmc** [comment]
Hmm that's not good.  If you can narrow down the cause of the performance problem, you should open a new issue about it.  I wasn't aware of any performance issues being reported in v1.22 yet.

**Comment by ADARSHAGNIHOTRI** [comment]
Any update on this? Is it fixed?

**Comment by jmagman** [comment]
The quick fix is in [v1.22.0-1.0.pre](https://flutter.dev/docs/development/tools/sdk/releases?tab=macos), which is currently available on the [dev channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#dev).

**Comment by m-j-g** [comment]
> The quick fix is in [v1.22.0-1.0.pre](https://flutter.dev/docs/development/tools/sdk/releases?tab=macos), which is currently available on the [dev channel](https://github.com/flutter/flutter/wiki/Flutter-build-release-channels#dev).

is this not also included in 1.22.0-12.1.pre?  I'm using this version from the beta channel (I'm assuming its a more recent version since the number is higher?) But I'm still getting "App pasted from.." message every time the app launches.

**Comment by justinmc** [comment]
@m-j-g Keep us posted if it ends up being something in the framework itself.

**Comment by m-j-g** [comment]
@justinmc confirmed that removing the flutter fire dynamic links dependency fixes the issue.  (of course now the apps functionality is compromised)

**Comment by justinmc** [comment]
Thanks.  Hopefully the package can be updated to fix it, I imagine more and more people are going to be using iOS14 and having the problem now.

**Comment by xster** [comment]
@justinmc can you summarize the remaining work here? It seems like we added the hasStrings handler on the engine for various platforms but haven't added it to the framework side in https://github.com/flutter/flutter/blob/master/packages/flutter/lib/src/services/clipboard.dart or consumed it in text_selection.dart yet right?

**Comment by justinmc** [comment]
That's right.  There are still a few more platforms to go, and once those are all done, then I can update the framework to use them.

iOS: https://github.com/flutter/engine/pull/19859
Android: https://github.com/flutter/engine/pull/20393
Mac: https://github.com/flutter/engine/pull/20531
Linux: https://github.com/flutter/engine/pull/21388
Windows: TODO
Fuchsia: TODO

Framework: https://github.com/flutter/flutter/pull/61774

**Comment by benPesso** [comment]
@justinmc @xster What is the expected impact on the `Clipboard` class and its behavior? Will there be a new `Clipboard.hasData()` method, for example, or will `Clipbaord.getData()` handle it internally and just return null if the clipboard is empty?

Side bar: Anyone else find it weird that when you paste in iOS14 (let's say, in Safari's address bar) the notification still pops up? It feels like Apple didn't bake this feature completely, as they could have easily created a proper interface for "legitimate" clipboard interactions, such a user purposefully pasting in an input field.

**Comment by justinmc** [comment]
Yes, there will be a `Clipboard.hasStrings()`.  See https://github.com/flutter/flutter/pull/61774.

I did find it unexpected that the notification popped up even when deliberately hitting the native paste button.  It may be a blessing in disguise for Flutter though, because Flutter doesn't use the native text selection menu.  I was worried that only Flutter apps would show the notification when hitting paste, but fortunately all apps do.

Maybe it's possible for apps to do sketchy stuff like effectively tap the paste button themselves, and that's why Apple did it?  Or to programmatically hit paste for accessibility purposes?

**Comment by benPesso** [comment]
Thanks, I saw the tickets. It just wasn't clear from the discussion here if the fix will be "under the hood" or not.

Personally, I think it makes more sense to handle this within the `Clipboard.getData` method. I don't see the point in pushing the extra check forced by Apple onto the developer. It doesn't add any value to the current workflow in Flutter.

**Comment by justinmc** [comment]
I don't think I could do this inside of getData.  hasStrings only tells you if there is a pasteable string on the clipboard, but it doesn't tell you what that string is.  That's why it doesn't trigger Apple's notification.

If the user opens a Flutter app with a text field, Flutter will check hasStrings to see if it can show the paste button.  If the user pastes, Flutter will use getData and the notification will be shown.  This all happens automatically from the perspective of the Flutter app developer.  If the app developer wants some special clipboard functionality, then they case use hasStrings or getData as they see fit.

**Comment by benPesso** [comment]
I think we're talking about two different flows here. I'm referring to developers using the `getData` method directly, to show a prompt to use the clipboard's content, for example (similar to what Chrome's address bar does).

So in the context of that, I would image `getData to look something like this:
```dart
    if (format.startsWith('text\/')) {
      final bool hasString =
          await SystemChannels.platform.invokeMethod('Clipboard.hasString');
      if (hasString != true) return null;
    }

    final Map<String, dynamic>? result =
        await SystemChannels.platform.invokeMethod(
      'Clipboard.getData',
      format,
    );
    if (result == null) return null;
    
    return ClipboardData(text: result['text'] as String?);
  }
```

This way, the new check is handled internally by Flutter, and developers don't have to add platform-specific logic to their code.

P.S.
You could potentially add a condition for the platform, but I think it's better for the platform channel to implement the hasString on Android/Fuchsia/etc., rather than add more conditions in the code.

**Comment by benPesso** [comment]
BTW, I think it might be a good idea to add type-specific methods to Flutter's `Clipboard` utility:

- `Clipboard.getImages`
Used most commonly in chat apps. Should implement `UIPasteboard` and `hasImages`. ([Docs](https://developer.apple.com/documentation/uikit/uipasteboard/1829411-hasimages)) 
- `Clipboard.getNumbers`
Used most commonly for auto-verifications. Should implement `UIPasteboard` and `detectPatterns`. ([Docs](https://developer.apple.com/documentation/uikit/uipasteboard/detectionpattern)) 
- `Clipboard.getStrings`
Used most commonly in journaling/editing apps. Should implement `UIPasteboard` and `hasStrings`. ([Docs](https://developer.apple.com/documentation/uikit/uipasteboard/1829416-hasstrings))
- `Clipboard.getUrls`
Used most commonly for deep-linking and auto-population of URLs. Should implement `UIPasteboard` and `hasURLs`. ([Docs](https://developer.apple.com/documentation/uikit/uipasteboard/1829410-hasurls))

Adding these will drastically reduce the number of irrelevant notifications the user will see in a Flutter app.

Actually, maybe this comment should be converted to a separate feature task...?

**Comment by justinmc** [comment]
@benPesso Yes, could you open a separate issue and tag me in it?  I wasn't aware of those other UIPasteboard methods but it looks like they could be useful.

We'll still need a separate Clipboard.hasStrings for the hiding and showing of the paste button, but it might make sense to do that hasStrings check inside of getData if it does avoid showing the notification when empty.

**Comment by benPesso** [comment]
@justinmc Done.

As for the `getData` method... As I wrote above, I think it should definitely handle it internally. I don't think it'll require a separate `Clipboard.hasStrings` method, though. I would make that check on the iOS side, when the platform method is invoked, and not on the Dart side.

**Comment by justinmc** [comment]
Flutter needs to use hasStrings to decide whether or not to show the paste button, before the user actually tries to paste anything.  In the case where the clipboard has a pasteable string on it, we don't want to ask what that string is and trigger iOS's notification before we need it.  Instead, we can use hasStrings without triggering the notification.

**Comment by benPesso** [comment]
Yeah, I understand, but maybe you don't understand what I'm saying...

Currently, Flutter is using the `Clipboard.getData` method to see if it should show the button or not. Since it's passing a format type to the call, the platform code knows if Flutter is looking for a string or not. If you add a call in the platform level to use the `hasStrings` API and return `null` if no string exists, there will be no trigger of the iOS notification, because you will make a `return` before accessing the iOS API that triggers it.

I tried finding the platform code so I can show you an actual example, but I couldn't find where it was. If you show me where to find it, I'll add an example to the discussion.

**Comment by justinmc** [comment]
Ok that makes sense as a way to avoid the paste notification when the clipboard is empty.  Is this the platform code you were looking for? https://github.com/flutter/engine/blob/ed47fc92ee52a2e53d426e8587130cc1c84d4718/shell/platform/darwin/ios/framework/Source/FlutterPlatformPlugin.mm#L235

**Comment by benPesso** [comment]
@justinmc Yeah, that's what I was looking for. Thanks.

So, in `getClipboardData` you can just add the `clipboardHasStrings` you've made as another condition to the `if (!format || ([format isEqualToString:@(kTextPlainFormat)]))` condition. And that's it. No more notifications on empty clipboards.

**Comment by justinmc** [comment]
Would you mind opening another issue for that as well?  Or even a pull request, if you've got time and you think you'd be able to write a test for that as well.

**Comment by mfouadegy** [comment]
any updates for this issue ?

**Comment by justinmc** [comment]
This is fixed for users and you should not ever see an unsolicited clipboard prompt in iOS 14 now.  However, a deeper fix regarding enabling and disabling the paste button is still in development.

**Comment by billymahmood** [comment]
I have the most up to date version of Flutter, and I have updated my app, however I am still getting this "copied from clipboard" notification on the launch of my app (when the app is completely closed and then opened).


```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 1.22.3, on Mac OS X 10.15.7 19H2, locale en-GB)
 
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.3)
[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
[✓] Android Studio (version 4.0)
[✓] Connected device (1 available)

• No issues found!
```


Packages my app uses
```
cupertino_icons: ^1.0.0
  http: ^0.12.2
  flutter_secure_storage: ^3.3.5
  connectivity: ^2.0.1
  device_info: ^1.0.0
  sqflite: ^1.3.1+2
  path_provider: ^1.6.24
  fluttertoast: ^7.1.1
  sprintf: ^5.0.0
  url_launcher: ^5.7.2
  firebase_messaging: ^7.0.3
  firebase_dynamic_links: ^0.6.0+2
  get_it: ^5.0.1
  syncfusion_flutter_gauges: ^18.3.40
  image_picker: ^0.6.7+12
  cached_network_image: ^2.3.3
  file_picker: ^2.0.10
  flutter_app_badger: ^1.1.2
```

**Comment by justinmc** [comment]
@billymahmood Something in your app (or a package) is probably accessing the clipboard.  Can you narrow it down and find out what it is?

**Comment by benPesso** [comment]
@billymahmood @justinmc It's most likely `firebase_dynamic_links`, which uses the clipboard to "reload" the dynamic link that was created when the user was redirected. It's a known trick of "sharing" data with an app from the browser, while not having a direct flow from browser to app.
For example: When a user doesn't have the app and first needs to download it, it would set the clipboard in the browser, then once the app is launched, it checks to see if there's a dynamic link to load.

**Comment by justinmc** [comment]
Thanks @benPesso!  I wasn't aware.  Sounds like something that's not fixable, but I've opened an issue in the flutterfire repo in case there's anything they can do: https://github.com/FirebaseExtended/flutterfire/issues/4051

**Comment by benPesso** [comment]
@justinmc I got a bit lost with all the multiple related issues. Is there any progress with adding the `hasStrings()` check to the Flutter `Clipboard` utility (or the underlying `Clipboard.getData` platform channel?

**Comment by justinmc** [comment]
No, I haven't worked on it.

**Comment by pcsosinski** [comment]
@justinmc can you update the milestone on this issue?

**Comment by justinmc** [comment]
Whoops, done thanks!

**Comment by yacineblr** [comment]
any updates for this issue ?

**Comment by justinmc** [comment]
This should be the final PR to fix this issue the right way, unless I've missed something: https://github.com/flutter/flutter/pull/87678
